manifests/server/grant.pp

```diff
@@ -187,37 +239,58 @@
       $schema = $object_name
 
       # Again there seems to be no easy way in plain SQL to check if ALL
-      # PRIVILEGES are granted on a table. By convention we use INSERT
-      # here to represent ALL PRIVILEGES (truly terrible).
-      $custom_privilege = $_privilege ? {
-        'ALL'            => 'INSERT',
-        'ALL PRIVILEGES' => 'INSERT',
-        default          => $_privilege,
+      # PRIVILEGES are granted on a table.
+      # There are currently 7 possible priviliges:
+      # ('SELECT','UPDATE','INSERT','DELETE','TRIGGER','REFERENCES','TRUNCATE')
+      # This list is consistant from Postgresql 8.0
+      #
+      # There are 4 cases to cover, each with it's own distinct unless clause:
+      #    grant ALL
+      #    grant SELECT (or INSERT or DELETE ...)
+      #    revoke ALL
+      #    revoke SELECT (or INSERT or DELETE ...)
+
+      if $ensure == 'present' {
+        if $_privilege == 'ALL' or $_privilege == 'ALL PRIVILEGES' {
+          # GRANT ALL
+          $custom_unless = "SELECT 1 FROM
+             ( SELECT tablename FROM pg_catalog.pg_tables
+               WHERE schemaname = '${schema}' EXCEPT
+                SELECT table_name FROM
+                (SELECT table_name,count(privilege_type) FROM information_schema.role_table_grants
+                  WHERE grantee = '${role}' AND table_schema = '${schema}'
+                  AND privilege_type IN ('SELECT','UPDATE','INSERT','DELETE','TRIGGER','REFERENCES','TRUNCATE')
+                  GROUP BY table_name
+                ) AS P WHERE P.count >= 7
+             ) AS TBLS HAVING TBLS.count=0"
```

Comment:
I see the reason to keep the logic reversed.  We can encapsulate whose thing again inside `NOT EXISTS` to keep it that way, so:

```sql
SELECT 1
WHERE NOT EXISTS (
    SELECT 1
    FROM pg_catalog.pg_tables AS t,
        (VALUES ('SELECT'), ('UPDATE'), ('INSERT'), ('DELETE'), ('TRIGGER'), ('REFERENCES'), ('TRUNCATE')) AS p(privilege_type)
    WHERE t.schemaname = '${schema}'
        AND NOT EXISTS (
            SELECT 1
            FROM information_schema.role_table_grants AS g
            WHERE g.grantee = '${role}'
                AND g.table_schema = t.schemaname
                AND g.table_name = t.tablename
                AND g.privilege_type = p.privilege_type
        )
)
```